internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,117 @@
+	flushedBufferStartOffset int64
+	flushedBufferEndOffset   int64
+
+	// Holds the last offset to which data is written.
+	fileSize int64
 }
 
 // CreateInMemoryWriteBuffer creates a buffer with InMemoryWriteBuffer.buffer
 // set to nil. Memory is allocated to the buffer when the first write call comes.
 // This avoids unnecessarily bloating GCSFuse memory consumption.
 //
-// To allocate memory to the buffer, use WriteBuffer.InitializeBuffer.
+// To allocate memory to the buffer, use WriteBuffer.Initialize.
 func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
 	b := &InMemoryWriteBuffer{}
 	// TODO: set mtime attribute.
 	return b
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+func (b *InMemoryWriteBuffer) Initialize(sizeInMB int) {
+	if ChunkSize == 0 {
+		ChunkSize = int64(sizeInMB * MiB)
+	}
+	if b.currentBuffer == nil {
+		b.currentBuffer = make([]byte, 0, ChunkSize)
+		b.currentBufferEndOffset = ChunkSize
+	}
+	if b.flushedBuffer == nil {
+		b.flushedBuffer = make([]byte, 0, ChunkSize)
+	}
+}
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, receivedContentStartOffset int64) error {
+	dataSize := int64(len(content))
+	receivedContentEndOffset := receivedContentStartOffset + dataSize
+
+	// If receivedContentStartOffset is not in the range of offsets stored in current
+	// buffer, trigger temp file flow for writes.
+	if receivedContentStartOffset < b.currentBufferStartOffset ||
+		b.currentBufferEndOffset < receivedContentStartOffset {
+		// TODO: finalise write and trigger temp file flow.
+		return fmt.Errorf(NonSequentialWriteError)
+	}
+
+	// If data can be completely written to the current buffer block, write it.
+	if receivedContentEndOffset <= b.currentBufferEndOffset {
+		return b.copyDataToBuffer(receivedContentStartOffset, receivedContentEndOffset, content)
+	} else {
+		// Below logic is written with the assumption that the data received in a
+		// single write call from the kernel will never exceed beyond buffer block size.
+		// This should always hold true because kernel writes never exceed 1 MiB size.
+
+		// Write the chunk of data that can be written to the current buffer block.
+		l := b.currentBufferEndOffset - receivedContentStartOffset
+		if err := b.copyDataToBuffer(receivedContentStartOffset, b.currentBufferEndOffset, content[:l]); err != nil {
+			return err
+		}
+
+		//TODO: get status of last write to GCS.
+		// If unsuccessful, trigger temp file flow.
+		// Else:
+		b.swapBuffers()
+		// TODO: trigger async upload to GCS for flushedBuffer.
+
+		// Update buffer offsets for next buffer block.
+		b.currentBufferStartOffset = b.currentBufferEndOffset
+		b.currentBufferEndOffset += ChunkSize
+		// Write the remaining data to currentBuffer.
+		return b.copyDataToBuffer(b.currentBufferStartOffset, receivedContentEndOffset, content[l:])
```

Comment:
This case will never happen as kernel write calls never exceed 1 MiB size and minimum buffer size that a user can configure is 1 MiB as well. So if we are advancing the chunk, we will always have the space to write the data.

Reply:
This is a generic utility, which in this case will get its data from kernel, so that assumption will never be tested in this use case. But as this is a public interface, it can really be used by any other caller. 

So, let's explicitly state that assumption in the function/interface-function header, and add check (in implementation that that case isn't supported) and test it (as a negative unit test).

Reply:
handled the case in copyDataToBuffer